File: src/examples/diksha/px4_sim/codegen/mex/run_sim/AeroMEst.c
/*
 * AeroMEst.c
 *
 * Code generation for function 'AeroMEst'
 *
 */

/* Include files */
#include "rt_nonfinite.h"
#include "run_sim.h"
#include "AeroMEst.h"
#include "mpower.h"
#include "norm.h"
#include "run_sim_data.h"

/* Function Definitions */
void AeroMEst(const real_T eul[3], const real_T x_dot[3], const real_T omega[3],
              const real_T Fa[3], real_T Moment_aero[3])
{
  int32_T i;
  real_T V;
  real_T dv14[9];
  real_T a[3];
  int32_T i3;
  real_T xw_dot[3];
  real_T alpha;
  static const int8_T b_a[9] = { 0, 0, -1, 0, 1, 0, 1, 0, 0 };

  real_T My_w_ac;
  covrtLogFcn(&emlrtCoverageInstance, 6U, 0);
  covrtLogBasicBlock(&emlrtCoverageInstance, 6U, 0);

  /* UNTITLED2 Summary of this function goes here */
  /*    Detailed explanation goes here */
  for (i = 0; i < 3; i++) {
    Moment_aero[i] = 0.0;
  }

  covrtLogFcn(&emlrtCoverageInstance, 4U, 0);
  covrtLogBasicBlock(&emlrtCoverageInstance, 4U, 0);
  V = norm(x_dot);
  if (covrtLogIf(&emlrtCoverageInstance, 6U, 0U, 0, V == 0.0)) {
    covrtLogBasicBlock(&emlrtCoverageInstance, 6U, 1);
  } else {
    covrtLogBasicBlock(&emlrtCoverageInstance, 6U, 2);
    dv14[0] = muDoubleScalarCos(eul[1]) * muDoubleScalarCos(eul[0]);
    dv14[1] = -muDoubleScalarCos(eul[2]) * muDoubleScalarSin(eul[0]) +
      muDoubleScalarSin(eul[2]) * muDoubleScalarSin(eul[1]) * muDoubleScalarCos
      (eul[0]);
    dv14[2] = muDoubleScalarSin(eul[2]) * muDoubleScalarSin(eul[0]) +
      muDoubleScalarCos(eul[2]) * muDoubleScalarSin(eul[1]) * muDoubleScalarCos
      (eul[0]);
    dv14[3] = muDoubleScalarCos(eul[1]) * muDoubleScalarSin(eul[0]);
    dv14[4] = muDoubleScalarCos(eul[2]) * muDoubleScalarCos(eul[0]) +
      muDoubleScalarSin(eul[2]) * muDoubleScalarSin(eul[1]) * muDoubleScalarSin
      (eul[0]);
    dv14[5] = -muDoubleScalarSin(eul[2]) * muDoubleScalarCos(eul[0]) +
      muDoubleScalarCos(eul[2]) * muDoubleScalarSin(eul[1]) * muDoubleScalarSin
      (eul[0]);
    dv14[6] = -muDoubleScalarSin(eul[1]);
    dv14[7] = muDoubleScalarSin(eul[2]) * muDoubleScalarCos(eul[1]);
    dv14[8] = muDoubleScalarCos(eul[2]) * muDoubleScalarCos(eul[1]);
    for (i = 0; i < 3; i++) {
      a[i] = 0.0;
      for (i3 = 0; i3 < 3; i3++) {
        a[i] += dv14[i + 3 * i3] * x_dot[i3];
      }
    }

    for (i = 0; i < 3; i++) {
      xw_dot[i] = 0.0;
      for (i3 = 0; i3 < 3; i3++) {
        xw_dot[i] += (real_T)b_a[i + 3 * i3] * a[i3];
      }
    }

    if (covrtLogIf(&emlrtCoverageInstance, 6U, 0U, 1, xw_dot[0] != 0.0)) {
      covrtLogBasicBlock(&emlrtCoverageInstance, 6U, 3);
      alpha = muDoubleScalarAtan2(-xw_dot[2], xw_dot[0]);
    } else {
      covrtLogBasicBlock(&emlrtCoverageInstance, 6U, 4);
      alpha = 0.0;
    }

    covrtLogBasicBlock(&emlrtCoverageInstance, 6U, 5);
    mpower(V);
    for (i = 0; i < 3; i++) {
      a[i] = 0.0;
      for (i3 = 0; i3 < 3; i3++) {
        a[i] += (real_T)b_a[i + 3 * i3] * omega[i3];
      }
    }

    My_w_ac = 0.6125 * mpower(V) * 0.377 * 0.29648527679623088 * ((-0.0156 +
      0.995 * alpha) + -0.51 * a[1] * 0.29648527679623088 / (2.0 * V));
    mpower(V);

    /*  wing frame */
    /*  r = BQ.x_ac - BQ.x_cg; */
    /*  M_cg = M_ac + cross(r',(Rq2w*Fa)')'; */
    /*  L = -Fa(1); */
    /*  M_cg = M_ac + [0;r*L;0]; % wing frame */
    /*  Moment_aero = BQ.wing_n*Rq2w'*M_cg; % body frame */
    Moment_aero[0] = 0.0;
    Moment_aero[1] = 2.0 * My_w_ac + -0.0732 * Fa[0];
    Moment_aero[2] = 0.0;
  }
}

/* End of code generation (AeroMEst.c) */

File: src/examples/diksha/px4_sim/AeroMEst.m
function [ Moment_aero ] = AeroMEst( eul, x_dot,omega, Fa);
%UNTITLED2 Summary of this function goes here
%   Detailed explanation goes here
coder.inline('never');
BQ_x_ac = [0.0838;0;0];
BQ_x_cg = [0.157;0;0];
c_root=0.39;
c_tip=0.176;
t= c_tip/c_root;
BQ_c = c_root*(2/3)*((1+t+t^2)/(1+t));
BQ_b=2.29/2;
BQ_S=0.754/2;

BQ_Cm_0=-0.0156;
BQ_Cm_alpha=0.995;
BQ_Cm_q=-0.51;
BQ_Cl_beta=0;
BQ_Cl_p=-0.43;
BQ_Cl_r=0.29;
BQ_Cn_beta=0.0812;
BQ_Cn_p=-0.4044;
BQ_Cn_r=-0.05;

rho=1.225;
Moment_aero=[0;0;0];

R=eul2rotm(eul);
xb_dot=R'*x_dot;
V=norm(x_dot);

if (V==0) return;
end

Rq2w = [
    0 0 1;
    0 1 0;
    -1 0 0];
xw_dot = Rq2w*xb_dot;

if (xw_dot(1)~=0)
    alpha=atan2(-xw_dot(3),xw_dot(1));
    beta=atan2(xw_dot(2),xw_dot(1));
else
    alpha=0;
    beta=0;
end

omega_w = Rq2w*omega;
p_w = omega_w(1,1);
q_w = omega_w(2,1);
r_w = omega_w(3,1);

C_l = BQ_Cl_beta*beta + BQ_Cl_p*p_w*BQ_b/(2*V) + BQ_Cl_r*r_w*BQ_b/(2*V);
C_m = BQ_Cm_0 + BQ_Cm_alpha*alpha + BQ_Cm_q*q_w*BQ_c/(2*V);
C_n = BQ_Cn_beta*beta + BQ_Cn_p*p_w*BQ_b/(2*V) + BQ_Cn_r*r_w*BQ_b/(2*V);

Mx_w_ac = 0.5*rho*(V^2)*BQ_S*BQ_c*C_l ;
My_w_ac = 0.5*rho*(V^2)*BQ_S*BQ_c*C_m ;
Mz_w_ac = 0.5*rho*(V^2)*BQ_S*BQ_c*C_n ;

M_ac = [Mx_w_ac;My_w_ac;Mz_w_ac];% wing frame
% r = BQ.x_ac - BQ.x_cg;
% M_cg = M_ac + cross(r',(Rq2w*Fa)')';
r = BQ_x_ac(1) - BQ_x_cg(1);
% L = -Fa(1);
% M_cg = M_ac + [0;r*L;0]; % wing frame

% Moment_aero = BQ.wing_n*Rq2w'*M_cg; % body frame

Moment_aero = [0;2*My_w_ac+r*Fa(1);0];

end

